core/admin.py

```diff
@@ -65,11 +67,19 @@ def algorithm(self, obj):
     def dataset(self, obj):
         return obj.algorithm_job.dataset
 
+    # how to combine the function?
     def log_preview(self, obj):
+        maxlen = 10000
+        log_message = 'The log output is too large to display in the browser.\n'
+        log_message += ('Only the last %s characters are displayed.\n \n' % (maxlen))
         if obj.log:
-            log = '\n'.join(obj.log.readlines())
+            log = '\n'.join(obj.log.open('rt').readlines())
```

Comment:
This is inefficient and won't work if there is binary data in the log.  Rather, we want to do something like:
```
if obj.log:
  prefix_message = ''
  with obj.log.open('rb') as datafile:
    try:
      datafile.seek(-maxlen, os.SEEK_END)
    except OSError as exc:
      if exc.errno != 22:  
        raise  # reraise exceptions except for trying to seek before the beginning
      prefix_message = ('The output is too large to display in the browser.\n' +
       f'Only the last {maxlen} characters are displayed.\n\n')
    message = datafile.read().decode(errors='replace')
    return prefix_message + message
```